eio/tls_eio.ml

```diff
@@ -5,198 +5,143 @@ exception Tls_failure of Tls.Engine.failure
+            Eio.Mutex.unlock t.mutex ;
+            application_data
+          | `Eof -> raise End_of_file ;
+          | `Alert a -> raise (Tls_alert a)
+        end
+      | Error (failure, `Response resp) ->
+        write_t t resp ;
+        raise (Tls_failure failure)
+    with 
+    | ( End_of_file | Tls_alert _ | Tls_failure _ ) as ex ->
+      Eio.Mutex.unlock t.mutex ;
       raise ex
-    | exn ->
-      (match t.state with
-       | `Error _ | `Eof -> ()
-       | `Active _ -> t.state <- `Error exn) ;
-      raise exn
-
-  let write_t t cs =
-    try Flow.copy (Flow.cstruct_source [cs]) t.flow
-    with exn ->
-      (match t.state with
-       | `Error _ | `Eof -> ()
-       | `Active _ -> t.state <- `Error exn) ;
-      raise exn
-
-  let try_write_t t cs =
-    try write_t t cs
-    with _ -> Eio.Fiber.check ()      (* Error is in [t.state] *)
-
-  let rec read_react t =
-
-    let handle tls buf =
-      match Tls.Engine.handle_tls tls buf with
-      | Ok (state', `Response resp, `Data data) ->
-          let state' = match state' with
-            | `Ok tls  -> `Active tls
-            | `Eof     -> `Eof
-            | `Alert a -> `Error (Tls_alert a)
-          in
-          t.state <- state' ;
-          Option.iter (try_write_t t) resp;
-          data
-
-      | Error (alert, `Response resp) ->
-          t.state <- `Error (Tls_failure alert) ;
-          write_t t resp; read_react t
-    in
 
-    match t.state with
-    | `Error e  -> raise e
-    | `Eof      -> raise End_of_file
-    | `Active _ ->
-        let n = read_t t t.recv_buf in
-        match (t.state, n) with
-        | (`Active tls, n) -> handle tls (Cstruct.sub t.recv_buf 0 n)
-        | (`Error e, _)    -> raise e
-        | (`Eof, _)        -> raise End_of_file
-
-  let rec read t buf =
-
-    let writeout res =
-      let open Cstruct in
-      let rlen = length res in
-      let n    = min (length buf) rlen in
-      blit res 0 buf 0 n ;
+  let rec read t flow_buf : int =
+
+    let write_application_data data =
+      let data_len = Cstruct.length data in
+      let n        = min (Cstruct.length flow_buf) data_len in
+      Cstruct.blit data 0 flow_buf 0 n ;
       t.linger <-
-        (if n < rlen then Some (sub res n (rlen - n)) else None) ;
+        if n < data_len
+        then Cstruct.sub data n (data_len -n)
+        else Cstruct.empty ;
       n in
 
-    match t.linger with
-    | Some res -> writeout res
-    | None     ->
-        match read_react t with
-          | None     -> read t buf
-          | Some res -> writeout res
-
-  let writev t css =
-    match t.state with
-    | `Error err  -> raise err
-    | `Eof        -> raise End_of_file
-    | `Active tls ->
-        match Tls.Engine.send_application_data tls css with
-        | Some (tls, tlsdata) ->
-            ( t.state <- `Active tls ; write_t t tlsdata )
-        | None -> invalid_arg "tls: write: socket not ready"
-
-  let write t cs = writev t [cs]
-
-  (*
```

Comment:
This is no longer applicable since the only place we need to worry about interleaving read/write is when reading from the tls_eio flow.